drivers/regulator/regulator_axp192.c

```diff
@@ -0,0 +1,408 @@
+#define AXP192_REG_CHIP_ID           0x03
+#define AXP192_REG_BUCK123_LDO23_CTL 0x12
+#define AXP192_REG_BUCK1_OUT_VOL     0x26
+#define AXP192_REG_BUCK2_OUT_VOL     0x23
+#define AXP192_REG_BUCK3_OUT_VOL     0x27
+#define AXP192_REG_LDO23_OUT_VOL     0x28
+#define AXP192_REG_VBUS_IPS_CTL      0x30
+#define AXP192_REG_CHG_CTL           0x33
+#define AXP192_REG_PER_KEY_CTL       0x36
+#define AXP192_REG_ACIN_IN_VOL       0x56
+#define AXP192_REG_ACIN_IN_CUR       0x58
+#define AXP192_REG_VBUS_IN_VOL       0x5A
+#define AXP192_REG_VBUS_IN_CUR       0x5C
+#define AXP192_REG_INTER_TEMP        0x5E
+#define AXP192_REG_BAT_VOL           0x78
+#define AXP192_REG_BAT_CHG_CUR       0x7A
+#define AXP192_REG_BAT_DCHG_CUR      0x7C
+
+/* X-Powers AXP192 voltage sources */
+enum axp192_voltage_sources {
+	AXP192_SOURCE_BUCK1,
+	AXP192_SOURCE_BUCK2,
+	AXP192_SOURCE_BUCK3,
+	AXP192_SOURCE_LDO1,
+	AXP192_SOURCE_LDO2,
+	AXP192_SOURCE_LDO3
+};
+
+struct regulator_axp192_pconfig {
+	struct i2c_dt_spec i2c;
+};
+
+struct regulator_axp192_config {
+	struct regulator_common_config common;
+	const struct device *p;
+	uint8_t source;
+};
+
+struct regulator_axp192_data {
+	struct regulator_common_data data;
+};
+
+/* Linear range for output voltage */
+static const struct linear_range buck13_range = LINEAR_RANGE_INIT(700000, 25000, 0U, 112U);
+static const struct linear_range buck2_range = LINEAR_RANGE_INIT(700000, 25000, 0U, 63U);
+static const struct linear_range ldo23_range = LINEAR_RANGE_INIT(1800000, 100000, 0U, 15U);
+
+static int buck_output_enable(const struct device *dev, uint8_t chan, uint8_t enable)
+{
+	const struct regulator_axp192_config *config = dev->config;
+	const struct regulator_axp192_pconfig *pconfig = config->p->config;
+	uint8_t mask_bit[6] = {0, 4, 1, 7, 2, 3};
+
+	if (enable) {
+		return i2c_reg_update_byte_dt(&pconfig->i2c, AXP192_REG_BUCK123_LDO23_CTL,
+					      BIT(mask_bit[chan]), BIT(mask_bit[chan]));
+	} else {
+		return i2c_reg_update_byte_dt(&pconfig->i2c, AXP192_REG_BUCK123_LDO23_CTL,
+					      BIT(mask_bit[chan]), 0);
+	}
+}
+
+static int buck_set_voltage(const struct device *dev, uint8_t chan, int32_t min_uv, int32_t max_uv)
+{
+	const struct regulator_axp192_config *config = dev->config;
+	const struct regulator_axp192_pconfig *pconfig = config->p->config;
+	uint16_t idx;
+	int ret;
+
+	if (chan == AXP192_SOURCE_BUCK2) {
+		ret = linear_range_get_win_index(&buck2_range, min_uv, max_uv, &idx);
+	} else {
+		ret = linear_range_get_win_index(&buck13_range, min_uv, max_uv, &idx);
+	}
+
+	if (ret == -EINVAL) {
+		return ret;
+	}
+
+	switch (chan) {
+	case AXP192_SOURCE_BUCK1:
+		return i2c_reg_write_byte_dt(&pconfig->i2c, AXP192_REG_BUCK1_OUT_VOL, idx);
+	case AXP192_SOURCE_BUCK2:
+		return i2c_reg_write_byte_dt(&pconfig->i2c, AXP192_REG_BUCK2_OUT_VOL, idx);
+	case AXP192_SOURCE_BUCK3:
+		return i2c_reg_write_byte_dt(&pconfig->i2c, AXP192_REG_BUCK3_OUT_VOL, idx);
+	default:
+		return -ENODEV;
+	}
+}
```

Comment:
suggestion here: this function could take a register and an index, determined by the caller function (which already knows if it's BUCK1/2/3. This would avoid checking source/chan 3 times.